File: Linked list/Add Two Numbers II/Add Two Numbers II.java

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode reverse(ListNode node){
        ListNode prev = null;
        ListNode curr = node;
        ListNode next = null;
        while(curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        node = prev;
        return node;
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode first = reverse(l1);
        ListNode sec = reverse(l2);
        ListNode res = new ListNode(0);
        ListNode temp = res;
        int carry = 0;
        while(first != null && sec != null){
           int sum = first.val + sec.val + carry;
           first = first.next;
           sec = sec.next;
           carry = sum / 10;
           temp.next = new ListNode(sum % 10);
           temp = temp.next;
        }
        while(first != null){
            int sum = first.val + carry;
            carry = sum / 10;
            temp.next = new ListNode(sum % 10);
            temp = temp.next;
            first = first.next;
        }
        while(sec != null){
            int sum = sec.val + carry;
            carry = sum / 10;
            temp.next = new ListNode(sum % 10);
            temp = temp.next;
            sec = sec.next;
        }
        if(carry > 0){
            temp.next = new ListNode(carry);
            temp = temp.next;
        }
        res = reverse(res.next);
        return res;
    }
}

File: Linked list/Add Two Numbers II/Add Two Numbers II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* node){
        ListNode* prev = NULL;
        ListNode* curr = node;
        ListNode* next = NULL;
        while(curr){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        node = prev;
        return node;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
       stack <int> s1;
       stack <int> s2;
        stack <int> s3;
        int carry = 0 ;
       while(l1){
           s1.push(l1->val);
           l1 = l1->next;
       }
      while(l2){
           s2.push(l2->val);
           l2 = l2->next;
       }
      while(!s1.empty() && !s2.empty()){
            int sum = s1.top() + s2.top() + carry;
            s2.pop();
            s1.pop();
            carry = sum / 10;
            s3.push(sum % 10);
        }
      while(!s1.empty()){
          int sum = s1.top() + carry;
          s1.pop();
          carry = sum / 10;
          s3.push(sum % 10);
      }
          while(!s2.empty()){
          int sum =  s2.top() + carry;
          s2.pop();
          carry = sum / 10;
          s3.push(sum % 10);
      }
        if(carry){
            s3.push(carry);
        }
        ListNode* newHead = new ListNode(0);
        ListNode* curr = newHead;
        while(!s3.empty()){
            curr->next = new ListNode(s3.top());
            curr = curr->next;
            s3.pop();
        }
        return newHead->next;
    }
};
